vsphere/resource_vsphere_resource_pool.go

```diff
@@ -0,0 +1,255 @@
+	client := meta.(*VSphereClient).vimClient
+	rp, err := resourcepool.FromID(client, d.Id())
+	if err != nil {
+		return err
+	}
+	_ = d.Set("name", rp.Name())
+	rpProps, err := resourcepool.Properties(rp)
+	if err != nil {
+		return err
+	}
+	err = flattenResourcePoolConfigSpec(d, &rpProps.Config)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] %s: Read finishes successfully", resourceVSphereResourcePoolIDString(d))
+	return nil
+}
+
+func resourceVSphereResourcePoolUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: Beginning update", resourceVSphereResourcePoolIDString(d))
+	client := meta.(*VSphereClient).vimClient
+	rp, err := resourcepool.FromID(client, d.Id())
+	if err != nil {
+		return err
+	}
+	rpSpec := expandResourcePoolConfigSpec(d)
+	err = resourcepool.Update(rp, d.Get("name").(string), rpSpec)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] %s: Update finished successfully", resourceVSphereResourcePoolIDString(d))
+	return nil
+}
+
+func resourceVSphereResourcePoolDelete(d *schema.ResourceData, meta interface{}) error {
+	log.Printf("[DEBUG] %s: Beginning delete", resourceVSphereResourcePoolIDString(d))
+	client := meta.(*VSphereClient).vimClient
+	rp, err := resourcepool.FromID(client, d.Id())
+	if err != nil {
+		return err
+	}
+	err = resourcepool.Delete(rp)
+	if err != nil {
+		return err
+	}
+	log.Printf("[DEBUG] %s: Deleted successfully", resourceVSphereResourcePoolIDString(d))
+	return nil
+}
+
+// resourceVSphereResourcePoolIDString prints a friendly string for the
+// vsphere_virtual_machine resource.
+func resourceVSphereResourcePoolIDString(d structure.ResourceIDStringer) string {
+	return structure.ResourceIDString(d, "vsphere_resource_pool")
+}
+
+func flattenResourcePoolConfigSpec(d *schema.ResourceData, obj *types.ResourceConfigSpec) error {
+	err := flattenResourcePoolMemoryAllocation(d, obj.MemoryAllocation)
+	if err != nil {
+		return err
+	}
+	return flattenResourcePoolCPUAllocation(d, obj.CpuAllocation)
+}
+
+func flattenResourcePoolCPUAllocation(d *schema.ResourceData, obj types.ResourceAllocationInfo) error {
+	return structure.SetBatch(d, map[string]interface{}{
+		"cpu_reservation": obj.Reservation,
+		"cpu_expandable":  obj.ExpandableReservation,
+		"cpu_limit":       obj.Limit,
+		"cpu_shares":      obj.Shares.Shares,
+		"cpu_share_level": obj.Shares.Level,
+	})
+}
+
+func flattenResourcePoolMemoryAllocation(d *schema.ResourceData, obj types.ResourceAllocationInfo) error {
+	return structure.SetBatch(d, map[string]interface{}{
+		"memory_reservation": obj.Reservation,
+		"memory_expandable":  obj.ExpandableReservation,
+		"memory_limit":       obj.Limit,
+		"memory_shares":      obj.Shares.Shares,
+		"memory_share_level": obj.Shares.Level,
+	})
+}
+
+func expandResourcePoolConfigSpec(d *schema.ResourceData) *types.ResourceConfigSpec {
+	return &types.ResourceConfigSpec{
+		CpuAllocation:    *expandResourcePoolCPUAllocation(d),
```

Comment:
I don't think the indirection here is necessary - `expandResourcePoolCPUAllocation` et al will only be ever returning values for fields that are always required, so the referenced functions should be modified to not return pointers, and the indirection removed here.